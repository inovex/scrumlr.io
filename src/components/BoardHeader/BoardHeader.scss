@import "constants/style";

.board-header {
  position: absolute;
  left: 30px;
  right: 30px;
  top: $column__border-width;
  height: $header__height;
  display: flex;
  flex-direction: row;
  align-items: center;
}

.board-header__link {
  position: absolute;
  left: 0;
  padding: $padding--small $padding--large;
  border-radius: 16px;
  width: max-content;
  cursor: pointer;
  transition: background-color 0.2s ease-in-out;
  outline: none;
  background-color: transparent;
  border: none;

  &:focus-visible {
    box-shadow: 0 0 0 2px rgba($color-backlog-blue, 0.32);
  }

  &:hover {
    background-color: rgba($color-backlog-blue, 0.06);
  }
}

[theme="dark"] {
  .board-header__link {
    &:focus-visible {
      box-shadow: 0 0 0 2px rgba($color-planning-pink, 0.48);
    }

    &:hover {
      background-color: rgba($color-planning-pink, 0.15);
    }
  }
}

.board-header__users {
  position: absolute;
  right: 0;
  padding: $padding--small;

  display: flex;
  flex-direction: row;
  align-items: center;

  background: none;
  border: none;
  outline: none;
}

.board-header_name-and-settings {
  position: absolute;
  left: 50%;
  transform: translateX(-50%);
  width: max-content;
  padding: $padding--small $padding--large;
  border-radius: 16px;
  transition: background-color 0.2s ease-in-out;

  outline: 0;
  border: 0;
  background: none;

  cursor: pointer;

  &:focus-visible {
    box-shadow: 0 0 0 2px rgba($color-backlog-blue, 0.32);
  }

  &:hover {
    background-color: rgba($color-backlog-blue, 0.06);
  }
}

[theme="dark"] {
  .board-header_name-and-settings {
    &:focus-visible {
      box-shadow: 0 0 0 2px rgba($color-planning-pink, 0.48);
    }

    &:hover {
      background-color: rgba($color-planning-pink, 0.15); // Ask designer for correct color
    }
  }
}

.board-header__access-policy-status {
  color: $color-middle-gray;
  font-size: $text-size--medium;
  font-weight: bold;
  letter-spacing: $letter-spacing--small;
  display: flex;
  align-items: center;
  width: max-content;
  margin: 0 auto;
}

.board-header__name-container {
  position: relative;
}

.board-header__name {
  display: inline-block;

  color: $color-black;
  font-size: 32px;
  font-weight: bold;
  letter-spacing: $letter-spacing--large;
  line-height: 40px;
  margin: 0;

  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
  max-width: 28vw;
}

[theme="dark"] {
  .board-header__name {
    color: $color-white;
  }
}

.board-header__access-policy-status-icon {
  display: inline;
  vertical-align: bottom;

  width: $icon--medium;
  height: $icon--medium;
  align-self: center;
  margin-right: 2px;
}

.board-header__access-policy-status-icon * {
  color: $color-middle-gray;
}

@media #{$tablet} {
  .board-header {
    left: 60px;
    right: 60px;
  }

  .board-header__name-container {
    min-height: 40px;
    min-width: 240px;
  }
}

@media #{$smartphone} {
  .board-header__users {
    right: 0;
  }

  .board-header__link {
    padding: $padding--small;
  }
}
