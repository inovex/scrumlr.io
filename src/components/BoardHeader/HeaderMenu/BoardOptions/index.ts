import {ShowAuthorOption} from "./ShowAuthorOption";
import {ShowHiddenColumnsOption} from "./ShowHiddenColumnsOption";
import {ShowOtherUsersNotesOption} from "./ShowOtherUsersNotesOption";
import {ShowAllBoardSettings} from "./ShowAllBoardSettings";

export const BoardOption = {
  ShowAuthorOption,
  ShowHiddenColumnsOption,
  ShowOtherUsersNotesOption,
  ShowAllBoardSettings,
};
