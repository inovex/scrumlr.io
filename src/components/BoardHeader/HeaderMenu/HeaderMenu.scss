@import "constants/style";

$header-menu-color--light: $color-backlog-blue;
$header-menu-color--dark: $color-planning-pink;
$header-menu-width: 288px;
$header-menu__item-height: 48px;

.header-menu {
  width: 288px;
  border-radius: 16px;
  box-shadow: 0 16px 32px 0 rgba(0, 87, 255, 0.24);
  overflow: hidden;
  list-style-type: none;
  padding: 0;
  margin: 0;
}
[theme="dark"] {
  .header-menu {
    background-color: $header-menu-color--dark;
    box-shadow: 0 16px 32px 0 rgba(0, 0, 0, 0.2);
  }
}
@media #{$tablet} {
  .header-menu {
    margin-top: -20px;
  }
}
