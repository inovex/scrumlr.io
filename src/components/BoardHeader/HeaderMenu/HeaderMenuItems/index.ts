export * from "./Author";
export * from "./BoardHeaderSetting";
export * from "./Columns";
export * from "./Delete";
export * from "./Export";
export * from "./Note";
export * from "./QRCode";
