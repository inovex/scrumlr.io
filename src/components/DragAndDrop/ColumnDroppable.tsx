import {Collision, UniqueIdentifier, useDndContext, useDndMonitor} from "@dnd-kit/core";
import {SortableContext, useSortable} from "@dnd-kit/sortable";
import classNames from "classnames";
import {COMBINE_THRESHOLD, MOVE_THRESHOLD} from "constants/misc";
import {ReactNode} from "react";
import store from "store";
import {Actions} from "store/action";

type ColumnDroppableProps = {
  id: UniqueIdentifier;
  items: UniqueIdentifier[];
  className?: string;
  children: ReactNode;
};

export const ColumnDroppable = ({className, children, items, id}: ColumnDroppableProps) => {
  const {setDroppableNodeRef, attributes, listeners, active, over, isOver} = useSortable({id, data: {type: "column"}});
  const {collisions} = useDndContext();

  const isNote = (collision: Collision) => collision.data?.droppableContainer.data.current.type === "note";
  const isWithinCombineRange = (collision: Collision) => collision.data?.value >= COMBINE_THRESHOLD && collision.data?.value < MOVE_THRESHOLD;
  const combineCollision = collisions?.find((collision) => isNote(collision) && isWithinCombineRange(collision) && collision.id !== over?.id);
  const isOverCollision = over?.data.current?.type === "note" ? collisions?.at(0) : undefined;

  const isOverColumn = () => {
    if (combineCollision) return items.includes(combineCollision.id);
    if (isOverCollision) return items.includes(isOverCollision.id);
    return isOver;
  };

  useDndMonitor({
    onDragEnd: () => {
      if (!active) return;
      const topCollision = collisions?.at(0);

      if (topCollision && topCollision.id === id) store.dispatch(Actions.editNote(active.id.toString(), {position: {column: id.toString(), stack: undefined, rank: 0}}));
    },
  });

  return (
    <SortableContext items={items}>
      <div ref={setDroppableNodeRef} className={classNames(className, {isOver: isOverColumn()})} {...attributes} {...listeners}>
        {children}
      </div>
    </SortableContext>
  );
};
