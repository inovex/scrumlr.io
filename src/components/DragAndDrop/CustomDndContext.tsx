import {
  Collision,
  CollisionDetection,
  defaultDropAnimationSideEffects,
  DndContext,
  DragOverEvent,
  DragOverlay,
  DragStartEvent,
  DropAnimation,
  MouseSensor,
  rectIntersection,
  TouchSensor,
  UniqueIdentifier,
  useSensor,
  useSensors,
} from "@dnd-kit/core";
import classNames from "classnames";
import {Note} from "components/Note";
import {getColorClassName} from "constants/colors";
import {MOVE_THRESHOLD} from "constants/misc";
import {ReactNode, useState} from "react";
import {useAppSelector} from "store";
import {Column} from "types/column";
import {isEqual} from "underscore";

type CustomDndContextProps = {
  children: ReactNode;
};

export const CustomDndContext = ({children}: CustomDndContextProps) => {
  const [maxCollision, setMaxCollision] = useState<Collision | undefined>();
  const [dragActive, setDragActive] = useState<{id: string; colorClassName: string} | undefined>();

  const {self, columns, notes} = useAppSelector((state) => ({self: state.participants?.self, columns: state.columns, notes: state.notes}), isEqual);

  const sensors = useSensors(useSensor(MouseSensor, {activationConstraint: {distance: 2}}), useSensor(TouchSensor, {activationConstraint: {delay: 200, tolerance: 5}}));
  const dropAnimation: DropAnimation = {
    sideEffects: defaultDropAnimationSideEffects({
      styles: {
        active: {
          opacity: "0.5",
        },
      },
    }),
  };

  const isNote = (collision: Collision) => collision.data?.droppableContainer.data.current.type === "note";

  const getColumn = (id?: UniqueIdentifier): Column | undefined => {
    if (!id) return undefined;
    const note = notes.find((n) => n.id === id);
    if (note) return columns.find((column) => column.id === note.position.column);
    return columns.find((column) => column.id === id);
  };

  const onDragStart = ({active}: DragStartEvent) => {
    if (!active) return;
    const activeColumnColor = getColumn(active.id)?.color;
    if (!activeColumnColor) return;
    setDragActive({id: active.id.toString(), colorClassName: getColorClassName(activeColumnColor)});
  };

  const onDragOver = ({collisions}: DragOverEvent) => {
    const overCollision = collisions?.at(0);
    if (!overCollision) return;
    if (isNote(overCollision)) setMaxCollision(overCollision);
  };

  const collisionDetectionWrapper =
    (collisionDetection: CollisionDetection) =>
    (...[args]: Parameters<typeof collisionDetection>) => {
      const collisions = collisionDetection(args);

      const overCollision = collisions.at(0);
      // if highest collision ratio is below threshold, place previous collision in front
      if (maxCollision && overCollision && isNote(overCollision) && overCollision.data?.value < MOVE_THRESHOLD) {
        // find last maxCollision in collisions
        const lastOver = collisions.findIndex((collision) => collision.id === maxCollision?.id);
        if (lastOver === -1) collisions.unshift({...maxCollision, data: {...maxCollision.data, value: 0}});
        else collisions.unshift(...collisions.splice(lastOver, 1));
      }

      return collisions;
    };

  return (
    <DndContext sensors={sensors} onDragStart={onDragStart} onDragOver={onDragOver} collisionDetection={collisionDetectionWrapper(rectIntersection)}>
      {children}

      <DragOverlay zIndex={1000} dropAnimation={dropAnimation} className={classNames("drag-overlay", dragActive?.colorClassName)}>
        {dragActive?.id && self ? <Note noteId={dragActive.id.toString()} viewer={self} /> : null}
      </DragOverlay>
    </DndContext>
  );
};
