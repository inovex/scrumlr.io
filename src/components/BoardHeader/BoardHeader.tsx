import "./BoardHeader.scss";
import lock from "assets/icon-lock.svg";
import BoardUsers from "components/BoardUsers/BoardUsers";
import HeaderLogo from "./HeaderLogo/HeaderLogo";

export interface BoardHeaderProps {
  boardstatus: string;
  name: string;
}

const BoardHeader = ({name, boardstatus}: BoardHeaderProps) => (
  <div className="board-header">
    <HeaderLogo />
    <div className="board-header__infos">
      <p className="board-header__status">{boardstatus}</p>
      <div className="board-header__title-block">
        <img className="board-header__status-image" src={lock} alt="Private Session" />
        <h1 className="board-header__title">{name}</h1>
      </div>
    </div>
    <div className="board-header__users">
      <BoardUsers />
    </div>
  </div>
);
export default BoardHeader;
