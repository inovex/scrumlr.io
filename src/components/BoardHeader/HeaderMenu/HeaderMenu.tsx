import {Portal} from "components/Portal";
import {useState} from "react";
import "./HeaderMenu.scss";
import {BoardOption} from "./BoardOptions";
import {BoardSettings} from "./BoardSettings";

type HeaderMenuProps = {
  open: boolean;
  onClose: () => void;
  currentUserIsModerator: boolean;
};

const HeaderMenu = (props: HeaderMenuProps) => {
  const [activeEditMode, setActiveEditMode] = useState(false);

  if (!props.open) {
    return null;
  }

  return (
    <Portal
      onClose={() => {
        setActiveEditMode(false);
        props.onClose();
      }}
    >
      <ul className="header-menu" onClick={(e) => e.stopPropagation()}>
        <BoardSettings activeEditMode={activeEditMode} currentUserIsModerator={props.currentUserIsModerator} setActiveEditMode={setActiveEditMode} />
        {props.currentUserIsModerator && (
          <>
            <BoardOption.ShowAuthorOption />
            <BoardOption.ShowOtherUsersNotesOption />
            <BoardOption.ShowHiddenColumnsOption />
          </>
        )}
        <BoardOption.ShowAllBoardSettings onClose={props.onClose} />
      </ul>
    </Portal>
  );
};

export {HeaderMenu};
