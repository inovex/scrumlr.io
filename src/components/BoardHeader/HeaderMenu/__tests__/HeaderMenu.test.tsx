import {Provider} from "react-redux";
import {screen} from "@testing-library/dom";
import {HeaderMenu} from "components/BoardHeader/HeaderMenu";
import {render} from "testUtils";
import getTestStore from "utils/test/getTestStore";
import {CustomDndContext} from "components/DragAndDrop/CustomDndContext";

Object.assign(navigator, {
  clipboard: {
    writeText: () => {},
  },
});

jest.mock("utils/export", () => ({
  ...jest.requireActual("utils/export"),
  exportAsJSON: jest.fn(),
  exportAsCSV: jest.fn(),
  exportAsCSVZip: jest.fn(),
}));

jest.mock("file-saver", () => ({saveAs: jest.fn()}));

const createHeaderMenu = (currentUserIsModerator: boolean) => {
  return (
    <Provider store={getTestStore()}>
      <HeaderMenu open onClose={() => undefined} currentUserIsModerator={currentUserIsModerator} />
    </Provider>
  );
};

describe("<HeaderMenu/>", () => {
  beforeEach(() => {
    const portal = global.document.createElement("div");
    portal.setAttribute("id", "portal");
    global.document.querySelector("body")!.appendChild(portal);
  });

  describe("should render correctly", () => {
    test("for moderator", () => {
      const {container} = render(createHeaderMenu(true), {container: global.document.querySelector("#portal")!});
      expect(container).toMatchSnapshot();
    });

    test("for participants", () => {
      const {container} = render(createHeaderMenu(false), {container: global.document.querySelector("#portal")!});
      expect(container).toMatchSnapshot();
    });

    describe("BoardOption.ShowAuthorOption", () => {
      test("Hide authors has correct label", () => {
        render(createHeaderMenu(true), {container: global.document.querySelector("#portal")!});
        expect(screen.getByTestId("author")).not.toBeNull();

        const label = screen.getByTestId("author")!.querySelector("span")!;
        expect(label).toHaveClass("board-option-button__label");
        expect(label.innerHTML).toEqual("Show authors of notes");
      });
    });

    describe("BoardOption.ShowNotesOfOtherUserOption", () => {
      test("Hide notes has correct label", () => {
        render(createHeaderMenu(true), {container: global.document.querySelector("#portal")!});
        expect(screen.getByTestId("note")).not.toBeNull();

        const label = screen.getByTestId("note")!.querySelector("span")!;
        expect(label).toHaveClass("board-option-button__label");
        expect(label.innerHTML).toEqual("Show notes of other users");
      });
    });

    describe("BoardOption.ShowHiddenColumnsOption", () => {
      test("Show columns has correct label", () => {
        render(createHeaderMenu(true), {container: global.document.querySelector("#portal")!});
        expect(screen.getByTestId("column")).not.toBeNull();

        const label = screen.getByTestId("column")!.querySelector("span")!;
        expect(label).toHaveClass("board-option-button__label");
        expect(label.innerHTML).toEqual("Show hidden columns for me");
      });
    });
  });
});
