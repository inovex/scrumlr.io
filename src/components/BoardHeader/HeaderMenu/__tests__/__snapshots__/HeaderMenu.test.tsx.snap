// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeaderMenu/> should render correctly for moderator 1`] = `
<div
  id="portal"
>
  <div
    class="portal"
    role="dialog"
  >
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
    <div
      data-focus-lock-disabled="false"
    >
      <div
        class="portal__frame accent-color__backlog-blue"
      >
        <div
          class="portal__content"
          role="dialog"
        >
          <ul
            class="header-menu"
          >
            <li
              class="board-settings"
            >
              <form>
                <input
                  class="board-settings__board-name"
                  disabled=""
                  placeholder="scrumlr.io"
                  value="test-board-name"
                />
                <button
                  class="board-settings__edit-button"
                  type="submit"
                >
                  edit
                </button>
              </form>
            </li>
            <li
              class="board-option"
              data-testid="author"
            >
              <button
                class="board-option-button"
              >
                <div
                  class="board-option-toggle"
                >
                  <div
                    class="board-option-toggle__switch board-option-toggle__switch--right"
                  />
                </div>
                <span
                  class="board-option-button__label"
                >
                  Show authors of notes
                </span>
              </button>
            </li>
            <li
              class="board-option"
              data-testid="note"
            >
              <button
                class="board-option-button"
              >
                <div
                  class="board-option-toggle"
                >
                  <div
                    class="board-option-toggle__switch board-option-toggle__switch--right"
                  />
                </div>
                <span
                  class="board-option-button__label"
                >
                  Show notes of other users
                </span>
              </button>
            </li>
            <li
              class="board-option"
              data-testid="column"
            >
              <button
                class="board-option-button"
              >
                <div
                  class="board-option-toggle"
                >
                  <div
                    class="board-option-toggle__switch board-option-toggle__switch--left"
                  />
                </div>
                <span
                  class="board-option-button__label"
                >
                  Show hidden columns for me
                </span>
              </button>
            </li>
            <li
              class="board-option"
              data-testid="allSettings"
            >
              <a
                class="board-option-link"
                href="/settings"
              >
                <p
                  class="board-option-link__label"
                >
                  Show all board settings
                </p>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
  </div>
</div>
`;

exports[`<HeaderMenu/> should render correctly for participants 1`] = `
<div
  id="portal"
>
  <div
    class="portal"
    role="dialog"
  >
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
    <div
      data-focus-lock-disabled="false"
    >
      <div
        class="portal__frame accent-color__backlog-blue"
      >
        <div
          class="portal__content"
          role="dialog"
        >
          <ul
            class="header-menu"
          >
            <li
              class="board-settings"
            >
              <form>
                <input
                  class="board-settings__board-name"
                  disabled=""
                  placeholder="scrumlr.io"
                  value="test-board-name"
                />
              </form>
            </li>
            <li
              class="board-option"
              data-testid="allSettings"
            >
              <a
                class="board-option-link"
                href="/settings"
              >
                <p
                  class="board-option-link__label"
                >
                  Show all board settings
                </p>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
  </div>
</div>
`;
